larpandoracontent/LArThreeDReco/LArTrackFragments/ClearTrackFragmentsTool.cc

```diff
@@ -52,52 +52,52 @@ bool ClearTrackFragmentsTool::FindTrackFragments(ThreeDTrackFragmentsAlgorithm *
-                throw StatusCodeException(STATUS_CODE_FAILURE);
+        // ATTN No longer guaranteed safe to use iterator after processing tensor element, hence caching results above
+        const Cluster *pFragmentCluster(nullptr);
+        this->ProcessTensorElement(pAlgorithm, overlapTensor, overlapResult, pFragmentCluster);
 
-            clustersToRemoveFromTensor.push_back(pClusterU);
-            clustersToRemoveFromTensor.push_back(pClusterV);
-            clustersToRemoveFromTensor.push_back(pClusterW);
-
-            ProtoParticle protoParticle;
-            ProtoParticleVector protoParticleVector;
-            protoParticle.m_clusterListU.push_back(pClusterU);
-            protoParticle.m_clusterListV.push_back(pClusterV);
-            protoParticle.m_clusterListW.push_back(pClusterW);
-            protoParticleVector.push_back(protoParticle);
-            particlesMade |= pAlgorithm->CreateThreeDParticles(protoParticleVector);
-        }
+        if (!pFragmentCluster)
+            throw StatusCodeException(STATUS_CODE_FAILURE);
 
-        clustersToRemoveFromTensor.insert(clustersToRemoveFromTensor.end(), modifiedClusters.begin(), modifiedClusters.end());
-        clustersToRemoveFromTensor.insert(clustersToRemoveFromTensor.end(), deletedClusters.begin(), deletedClusters.end());
+        if (TPC_VIEW_U == fragmentHitType) pClusterU = pFragmentCluster;
+        if (TPC_VIEW_V == fragmentHitType) pClusterV = pFragmentCluster;
+        if (TPC_VIEW_W == fragmentHitType) pClusterW = pFragmentCluster;
 
-        this->RebuildClusters(pAlgorithm, modifiedClusters, clustersToAddToTensor);
-        this->UpdateTensor(pAlgorithm, overlapTensor, clustersToRemoveFromTensor, clustersToAddToTensor);
+        if (!(pClusterU->IsAvailable() && pClusterV->IsAvailable() && pClusterW->IsAvailable()))
+            throw StatusCodeException(STATUS_CODE_FAILURE);
 
-        if (particlesMade)
-            return true;
+        // ATTN For safety, remove all clusters associated with this fragment particle from the tensor
+        ClusterList fragmentClusterList, affectedKeyClusters;
+        fragmentClusterList.push_back(pClusterU);
+        fragmentClusterList.push_back(pClusterV);
+        fragmentClusterList.push_back(pClusterW);
+        this->GetAffectedKeyClusters(overlapTensor, fragmentClusterList, affectedKeyClusters);
+
+        for (const Cluster *const pCluster : affectedKeyClusters)
+            pAlgorithm->UpdateUponDeletion(pCluster);
+
+        // Now make the particle
+        ProtoParticle protoParticle;
+        ProtoParticleVector protoParticleVector;
+        protoParticle.m_clusterListU.push_back(pClusterU);
+        protoParticle.m_clusterListV.push_back(pClusterV);
+        protoParticle.m_clusterListW.push_back(pClusterW);
+        protoParticleVector.push_back(protoParticle);
+        return pAlgorithm->CreateThreeDParticles(protoParticleVector);
```

Comment:
is it worth having the result of CreateThreeDParticles checked before the affected clusters go under UpdateUponDeletion? What would happen if for any reason the creation of the threed particle fails (and we return false at this point)?

Reply:
I thought about this, but the checks above in the tool essentially mean that false can _currently_ never be returned (we check cluster availabilities and indeed ensure that there are some clusters provided). If false is returned, the control is passed to the next tensor tool to have a go.

Even if we changed the implementation so that false could be returned, everything would be well defined: the worst that can happen is that some clusters could be taken out of the tensor, but (somehow) _not_ turned into a particle and would be unavailable for the next tensor tool.

With the current implementation, false cannot happen (exceptions could be raised inside CreateThreeDParticles, but this would terminate the algorithm entirely) and there also is no other tool to run next in the chain. I'm therefore minded to go with this as it has been well tested and is consistent with other tools used by other tensor-related algorithms.

[Being pedantic, we could raise an exception if CreateThreeDParticles doesn't return true. If it does return true, we'd then manually need to return true in the tool. Would probably want to review all tools from all tensor-related algorithms in a specific new piece of maintenance.]